### PR TITLE
refactor: split llmq/utils to Quorum Calculation and llmq/options

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
`llmq/utils` has simple util code that used all over code base and also have too heavy code for calculation quorums such as: `GetAllQuorumMembers`, `EnsureQuorumConnections` and other.

These helpers for calculation quorums are used only by evo/deterministicmns, evo/simplifiedmns and llmq/*  modules, but llmq/utils is included in many other modules for various trivial helpers.

## What was done?
Prior work:
 - https://github.com/dashpay/dash/pull/5753
 - #5486
 See also #4798

This PR remove all non-quorum calculation code from llmq/utils. Eventually it happens that easier to take everything out rather than move Quorum Calculation to new place atm:
 - new module llmq/options have a code related to various params, command line options, spork-related etc
 - llmq/utils is not included in various files which do not use any llmq/utils code
 - helper `BuildCommitmentHash` goes to llmq/commitment
 - helper `BuildSignHash` goes to llmq/signing
 - helper `GetLLMQParam` inlined since it's trivial (it has not been trivial when introduced ages ago)
 - removed dependency of `IsQuorumEnabled` on CQuorumManager which means `quorumManager` deglobalization is done for 90%

## How Has This Been Tested?
 - Run unit functional tests
 - updated circular dependencies `test/lint/lint-circular-dependencies.sh`
 - check that llmq/utils is not included without needs to calculate Quorums Members
```
$ grep -r include src/ 2> /dev/null | grep -v .Po: | grep -vE 'llmq/utils.(h|cpp)': | grep llmq/utils  
src/evo/mnauth.cpp:#include <llmq/utils.h>
src/evo/deterministicmns.cpp:#include <llmq/utils.h>
src/llmq/quorums.cpp:#include <llmq/utils.h>
src/llmq/blockprocessor.cpp:#include <llmq/utils.h>
src/llmq/commitment.cpp:#include <llmq/utils.h>
src/llmq/debug.cpp:#include <llmq/utils.h>
src/llmq/dkgsessionhandler.cpp:#include <llmq/utils.h>
src/llmq/dkgsession.cpp:#include <llmq/utils.h>
src/llmq/dkgsessionmgr.cpp:#include <llmq/utils.h>
src/rpc/quorums.cpp:#include <llmq/utils.h>
```

## Breaking Changes
N/A

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: src/llmq/blockprocessor.cpp [hunk 1]
```diff
@@ -346,6 +347,9 @@ bool CQuorumBlockProcessor::UndoBlock(const CBlock& block, gsl::not_null<const C
 
         // if a reorg happened, we should allow to mine this commitment later
         AddMineableCommitment(qc);
+
+        // if a reorg happened, we should allow to mine this commitment later
+        AddMineableCommitment(qc);
```
Comment (UdjinM6) [inline src/llmq/blockprocessor.cpp:352]:
```suggestion
```

File: src/llmq/blockprocessor.cpp [hunk 2]
```diff
@@ -365,10 +369,10 @@ bool CQuorumBlockProcessor::GetCommitmentsFromBlock(const CBlock& block, gsl::no
         if (tx->nType == TRANSACTION_QUORUM_COMMITMENT) {
             CFinalCommitmentTxPayload qc;
             if (!GetTxPayload(*tx, qc)) {
-                // should not happen as it was verified before processing the block
+               // should not happen as it was verified before processing the block
```
Comment (UdjinM6) [inline src/llmq/blockprocessor.cpp:372]: unrelated, pls revert

File: src/llmq/blockprocessor.cpp [hunk 3]
```diff
@@ -365,10 +369,10 @@ bool CQuorumBlockProcessor::GetCommitmentsFromBlock(const CBlock& block, gsl::no
...
                 LogPrint(BCLog::LLMQ, "CQuorumBlockProcessor::%s height=%d GetTxPayload fails\n", __func__, pindex->nHeight);
                 return state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "bad-qc-payload");
-            }
+           }
```
Comment (UdjinM6) [inline src/llmq/blockprocessor.cpp:375]: unrelated, pls revert

File: src/llmq/utils.cpp [hunk 4]
```diff
@@ -870,10 +870,7 @@ bool IsQuorumTypeEnabledInternal(Consensus::LLMQType llmqType, const CQuorumMana
             // fall through
         case Consensus::LLMQType::LLMQ_TEST_INSTANTSEND: {
             if (!fDIP0024IsActive) return true;
-
-            const bool fHaveDIP0024Quorums{optHaveDIP0024Quorums.value_or(!qman.ScanQuorums(
-                            consensusParams.llmqTypeDIP0024InstantSend, pindexPrev, 1).empty())};
-            return !fHaveDIP0024Quorums;
+            return false;
```
Comment (UdjinM6) [inline src/llmq/utils.cpp]: can be simplified to `return !fDIP0024IsActive;`

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — APPROVED** [review]
utACK

**Comment by PastaPastaPasta** [comment]
> to get merge after https://github.com/dashpay/dash/pull/5782 (due to its conflicts)

**Review by UdjinM6 — DISMISSED** [review]
re-utACK

**Comment by knst** [comment]
need guix image to test a theory with crash:
```
Thread 1 "dashd" received signal SIGSEGV, Segmentation fault.
....
#0  0x0000555555c45198 in llmq::CQuorumManager::ScanQuorums (this=this@entry=0x55555694a110, llmqType=Consensus::LLMQType::LLMQ_TEST_INSTANTSEND, 
    pindexStart=pindexStart@entry=0x555556d66880, nCountRequested=nCountRequested@entry=1) at llmq/quorums.cpp:495
#1  0x0000555555809089 in llmq::utils::IsQuorumTypeEnabledInternal (llmqType=<optimized out>, qman=..., pindexPrev=..., optDIP0024IsActive=..., optDIP0024IsActive@entry=..., 
    optHaveDIP0024Quorums=...) at llmq/utils.cpp:942
#2  0x00005555558091d6 in llmq::utils::IsQuorumTypeEnabled (llmqType=<optimized out>, qman=..., pindexPrev=...) at llmq/utils.cpp:923
#3  0x0000555555c45211 in llmq::CQuorumManager::ScanQuorums (this=this@entry=0x55555694a110, llmqType=Consensus::LLMQType::LLMQ_TEST_INSTANTSEND, 
    pindexStart=pindexStart@entry=0x555556d66880, nCountRequested=nCountRequested@entry=1) at llmq/quorums.cpp:496
#4  0x0000555555809089 in llmq::utils::IsQuorumTypeEnabledInternal (llmqType=<optimized out>, qman=..., pindexPrev=..., optDIP0024IsActive=..., optDIP0024IsActive@entry=..., 
    optHaveDIP0024Quorums=...) at llmq/utils.cpp:942
```

run as `dashd -regtest -llmqtestinstantsenddip0024=llmq_test_instantsend`

**Comment by PastaPastaPasta** [comment]
Guix Automation has began to build this PR tagged as v20.1.0-devpr5790.98457039. A new comment will be made when the image is pushed.

**Comment by PastaPastaPasta** [comment]
Guix Automation has completed; a release should be present here: https://github.com/dashpay/dash-dev-branches/releases/tag/v20.1.0-devpr5790.98457039. The image should be on dockerhub soon.

**Comment by knst** [comment]
> @[knst](https://github.com/knst) knst [force-pushed](https://github.com/dashpay/dash/compare/984570393bf9a94831d8b344e4214e19d0bf3641..45957d53bc311ae6fb2cf2aac7b0304c1cf6b299) the refactor-llmq_utils branch from[ 9845703 ](https://github.com/dashpay/dash/commit/984570393bf9a94831d8b344e4214e19d0bf3641)to 45957d5

rebase to top of develop + resolved conflicts

**Review by UdjinM6 — APPROVED** [review]
re-utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Pull request merged**